File: sdk/python/examples/end_to_end.py
"""
DAP end-to-end example

Shows a complete flow:
  1. Principal generates keys and issues a credential to an agent
  2. A mock service verifies the credential
  3. Agent "registers" and gets an access token
  4. Tampered credential is rejected
  5. Unknown principal is rejected

Run: python examples/end_to_end.py
"""

import asyncio

from dap_sdk import (
    AgentCard,
    DAPError,
    DAPPrincipal,
    DAPService,
    RegisterRequest,
    Verification,
)


async def main() -> None:
    print("-- DAP end-to-end example --\n")

    # -- Step 1: Principal generates keys --
    print("1. Generating principal keypair...")
    principal = DAPPrincipal.generate(
        principal_id="did:web:romashka.example",
        name="OOO Romashka",
    )
    print(f"   Principal: {principal.name} ({principal.principal_id})\n")

    # -- Step 2: Principal issues a credential to the agent --
    print("2. Issuing agent credential...")
    credential = principal.issue_credential(
        agent_id="urn:dap:agent:procurement-bot-v1",
        scope=["register", "read_data", "submit_forms"],
        principal_type="organization",
        purpose="Supplier procurement automation",
        contact_email="tech@romashka.example",
        expires_in="24h",
        verification=Verification(level="domain", method="did:web"),
    )
    print(f"   Credential (JWT): {credential[:60]}...\n")

    # -- Step 3: Set up a service that accepts DAP agents --
    print("3. Setting up service...")
    # Store known principal public keys
    known_keys = {principal.principal_id: principal.export_public_key()}

    service = DAPService(
        name="Supplier Portal",
        base_url="https://supplier-portal.example",
        resolve_public_key=lambda pid: _resolve(known_keys, pid),
        min_verification_level="self",
    )
    print(f"   Service ready: {service.discovery_document()['service_info']['name']}\n")

    # -- Step 4: Agent registers at the service --
    print("4. Agent registers at service...")
    result = await service.handle_register(
        RegisterRequest(
            credential=credential,
            agent_card=AgentCard(
                name="Procurement Bot v1",
                description="Supplier procurement agent for OOO Romashka",
                version="1.0.0",
                capabilities=["search", "compare", "request_quote"],
            ),
        )
    )

    print("   Registered!")
    print(f"   Status:          {result.status}")
    print(f"   Account ID:      {result.account_id}")
    print(f"   Principal:       {result.principal_display}")
    print(f"   Granted scope:   {', '.join(result.granted_scope)}")
    print(f"   Access token:    {result.access_token[:40]}...\n")

    # -- Step 5: Tampered credential should be rejected --
    print("5. Tampered credential (should fail)...")
    tampered = credential[:-10] + "TAMPERED00"
    try:
        await service.handle_register(
            RegisterRequest(
                credential=tampered,
                agent_card=AgentCard(name="Evil Bot"),
            )
        )
        print("   FAIL: Should have been rejected!")
    except DAPError as e:
        print(f"   Correctly rejected: {e.error}")

    # -- Step 6: Unknown principal should be rejected --
    print("\n6. Unknown principal (should fail)...")
    stranger = DAPPrincipal.generate(principal_id="did:web:evil.example", name="Evil Corp")
    stranger_credential = stranger.issue_credential(
        agent_id="urn:dap:agent:spy",
        scope=["register"],
    )
    try:
        await service.handle_register(
            RegisterRequest(
                credential=stranger_credential,
                agent_card=AgentCard(name="Spy Bot"),
            )
        )
        print("   FAIL: Should have been rejected!")
    except DAPError as e:
        print(f"   Correctly rejected: {e.error}")

    print("\n-- Done --")


def _resolve(known_keys: dict, principal_id: str):
    """Simple key resolver for the example."""
    if principal_id not in known_keys:
        raise KeyError(f"Unknown principal: {principal_id}")
    return known_keys[principal_id]


if __name__ == "__main__":
    asyncio.run(main())

File: sdk/typescript/examples/end-to-end.ts
/**
 * DAP end-to-end example
 *
 * Shows two flows:
 *   A. Self-issued: principal signs its own credential
 *   B. Delegated issuer: a trusted third party signs the credential on behalf of a principal
 *
 * Run: npx tsx examples/end-to-end.ts
 */

import { DAPPrincipal, DAPService } from "../src/index.js";
import { importJWK } from "jose";

async function main() {
  console.log("== DAP end-to-end example ==\n");

  // ═══════════════════════════════════════════════════════════════════════════
  // Flow A: Self-issued credential (organization)
  // ═══════════════════════════════════════════════════════════════════════════

  console.log("--- Flow A: Self-issued credential ---\n");

  // Step 1: Principal generates keys
  console.log("1. Generating principal keypair...");
  const principal = await DAPPrincipal.generate({
    id: "did:web:acme-corp.example",
    name: "Acme Inc",
  });
  console.log(`   Principal: ${principal.name} (${principal.id})\n`);

  // Step 2: Issue credential (self-issued, no issuer_id)
  console.log("2. Issuing agent credential (self-issued)...");
  const credential = await principal.issueCredential({
    agentId: "urn:dap:agent:procurement-bot-v1",
    scope: ["register", "read_data", "submit_forms"],
    principalType: "organization",
    purpose: "Supplier procurement automation",
    contactEmail: "tech@acme-corp.example",
    expiresIn: "24h",
    verification: { level: "domain", method: "did:web" },
  });
  console.log("   Credential (JWT):", credential.slice(0, 60) + "...\n");

  // Step 3: Set up a service
  console.log("3. Setting up service...");
  const principalPublicJWK = await principal.exportPublicKeyJWK();

  const service = new DAPService({
    name: "Supplier Portal",
    baseUrl: "https://supplier-portal.example",
    minVerificationLevel: "self",
    acceptedPrincipalSchemes: ["did:web", "https", "mailto"],
    resolvePublicKey: async (id) => {
      if (id === principal.id) return importJWK(principalPublicJWK, "EdDSA");
      throw new Error(`Unknown id: ${id}`);
    },
  });
  console.log("   Service ready:", service.discoveryDocument().service_info.name, "\n");

  // Step 4: Register
  console.log("4. Agent registers at service...");
  const result = await service.handleRegister({
    credential,
    agent_card: {
      name: "Procurement Bot v1",
      description: "Supplier procurement agent for Acme Inc",
      version: "1.0.0",
      capabilities: ["search", "compare", "request_quote"],
    },
  });

  console.log("   Registered!");
  console.log("   Status:         ", result.status);
  console.log("   Account ID:     ", result.account_id);
  console.log("   Principal:      ", result.principal_display);
  console.log("   Granted scope:  ", result.granted_scope.join(", "));
  console.log("   Access token:   ", result.access_token.slice(0, 40) + "...\n");

  // Step 5: Tampered credential should be rejected
  console.log("5. Tampered credential (should fail)...");
  const tampered = credential.slice(0, -10) + "TAMPERED00";
  try {
    await service.handleRegister({ credential: tampered, agent_card: { name: "Evil Bot" } });
    console.log("   FAIL: Should have been rejected!");
  } catch (e: unknown) {
    const err = e as Error & { dapError?: { error: string } };
    console.log(`   Correctly rejected: ${err.dapError?.error ?? err.message}`);
  }

  // Step 6: Unknown principal should be rejected
  console.log("\n6. Unknown principal (should fail)...");
  const stranger = await DAPPrincipal.generate({ id: "did:web:evil.example", name: "Evil Corp" });
  const strangerCredential = await stranger.issueCredential({
    agentId: "urn:dap:agent:spy",
    scope: ["register"],
  });
  try {
    await service.handleRegister({ credential: strangerCredential, agent_card: { name: "Spy Bot" } });
    console.log("   FAIL: Should have been rejected!");
  } catch (e: unknown) {
    const err = e as Error & { dapError?: { error: string } };
    console.log(`   Correctly rejected: ${err.dapError?.error ?? err.message}`);
  }

  // ═══════════════════════════════════════════════════════════════════════════
  // Flow B: Delegated issuer (individual via third-party issuer)
  // ═══════════════════════════════════════════════════════════════════════════

  console.log("\n\n--- Flow B: Delegated issuer credential ---\n");

  // Step 7: Issuer generates keys (e.g. a DAP Wallet provider)
  console.log("7. Generating issuer keypair (DAP Wallet)...");
  const issuer = await DAPPrincipal.generate({
    id: "https://dap-wallet.example",
    name: "DAP Wallet",
  });
  const issuerPublicJWK = await issuer.exportPublicKeyJWK();
  console.log(`   Issuer: ${issuer.name} (${issuer.id})\n`);

  // Step 8: Issuer issues a credential on behalf of a human principal.
  // The issuer signs with its own key, sets issuer_id to itself,
  // and sets principal_id/principal_name to the actual user.
  console.log("8. Issuer issues credential for individual principal...");
  const delegatedCredential = await issuer.issueCredential({
    agentId: "urn:dap:agent:booking-bot-v1",
    scope: ["register", "search", "book"],
    principalType: "individual",
    // Override principal identity to the actual user (not the issuer)
    principalId: "https://accounts.google.com/user/12345",
    principalName: "James Wilson",
    issuerId: "https://dap-wallet.example",
    purpose: "Restaurant search and booking",
    contactEmail: "james.w@example.com",
    expiresIn: "12h",
    verification: {
      level: "oauth",
      method: "google_oauth2",
      verified_at: Math.floor(Date.now() / 1000),
      verified_by: "https://dap-wallet.example",
    },
  });
  console.log("   Credential (JWT):", delegatedCredential.slice(0, 60) + "...\n");

  // Step 9: Set up a service that requires oauth-level verification and trusts the issuer
  console.log("9. Setting up booking service (requires oauth, trusts dap-wallet.example)...");
  const bookingService = new DAPService({
    name: "Restaurant Booking",
    baseUrl: "https://restaurant-booking.example",
    minVerificationLevel: "oauth",
    trustedIssuers: ["https://dap-wallet.example"],
    acceptedPrincipalSchemes: ["did:web", "https", "mailto"],
    resolvePublicKey: async (id) => {
      if (id === principal.id) return importJWK(principalPublicJWK, "EdDSA");
      if (id === "https://dap-wallet.example") return importJWK(issuerPublicJWK, "EdDSA");
      throw new Error(`Unknown id: ${id}`);
    },
  });
  const disc = bookingService.discoveryDocument();
  console.log("   Service:", disc.service_info.name);
  console.log("   min_verification_level:", disc.requirements.min_verification_level);
  console.log("   trusted_issuers:", disc.requirements.trusted_issuers, "\n");

  // Step 10: Register with delegated credential -- should succeed
  console.log("10. Agent registers with delegated credential...");
  const delegatedResult = await bookingService.handleRegister({
    credential: delegatedCredential,
    agent_card: {
      name: "Booking Bot v1",
      description: "Restaurant booking agent for James Wilson",
      version: "1.0.0",
      capabilities: ["search", "book"],
    },
  });

  console.log("   Registered!");
  console.log("   Status:         ", delegatedResult.status);
  console.log("   Account ID:     ", delegatedResult.account_id);
  console.log("   Principal:      ", delegatedResult.principal_display);
  console.log("   Granted scope:  ", delegatedResult.granted_scope.join(", "));
  console.log("   Access token:   ", delegatedResult.access_token.slice(0, 40) + "...\n");

  // Step 11: Untrusted issuer should be rejected
  console.log("11. Untrusted issuer (should fail)...");
  const rogueIssuer = await DAPPrincipal.generate({
    id: "https://rogue-issuer.example",
    name: "Rogue Issuer",
  });
  const rogueCredential = await rogueIssuer.issueCredential({
    agentId: "urn:dap:agent:rogue-bot",
    scope: ["register"],
    principalId: "https://accounts.google.com/user/99999",
    principalName: "Hacker",
    issuerId: "https://rogue-issuer.example",
    verification: { level: "oauth", method: "custom_oauth" },
  });
  try {
    await bookingService.handleRegister({ credential: rogueCredential, agent_card: { name: "Rogue Bot" } });
    console.log("   FAIL: Should have been rejected!");
  } catch (e: unknown) {
    const err = e as Error & { dapError?: { error: string } };
    console.log(`   Correctly rejected: ${err.dapError?.error ?? err.message}`);
  }

  // Step 12: Insufficient verification level should be rejected
  console.log("\n12. Insufficient verification level (self < oauth, should fail)...");
  const lowVerifCredential = await principal.issueCredential({
    agentId: "urn:dap:agent:low-verif-bot",
    scope: ["register"],
    // No verification field => defaults to "self", below "oauth" minimum
  });
  try {
    await bookingService.handleRegister({ credential: lowVerifCredential, agent_card: { name: "Low Verif Bot" } });
    console.log("   FAIL: Should have been rejected!");
  } catch (e: unknown) {
    const err = e as Error & { dapError?: { error: string } };
    console.log(`   Correctly rejected: ${err.dapError?.error ?? err.message}`);
  }

  // Step 13: Unsupported principal scheme should be rejected
  console.log("\n13. Unsupported principal scheme (tel: not in accepted list, should fail)...");
  const telPrincipal = await DAPPrincipal.generate({
    id: "tel:+79001234567",
    name: "Anna",
  });
  const telCredential = await telPrincipal.issueCredential({
    agentId: "urn:dap:agent:tel-bot",
    scope: ["register"],
    principalType: "individual",
    verification: { level: "phone", method: "sms" },
  });
  try {
    await bookingService.handleRegister({ credential: telCredential, agent_card: { name: "Tel Bot" } });
    console.log("   FAIL: Should have been rejected!");
  } catch (e: unknown) {
    const err = e as Error & { dapError?: { error: string } };
    console.log(`   Correctly rejected: ${err.dapError?.error ?? err.message}`);
  }

  console.log("\n== Done ==");
}

main().catch(console.error);
